### DISCUSSION TITLE
Serial to MQTT bridge

### DISCUSSION THREAD

**Discussion by radudorurus** [arendst/Tasmota#18636] [Support request]
Hello.

Please guide me to the right place where to find the documentation for implementing serial to MQTT bridge. I hope I named it correctly.
I've searched so far and I found incomplete information or something not suitable for what I want to do.

The project is simple.
I have an Arduino Nano standalone automation for the heating system (that includes a boiler, DHW tank, heat buffer tank, few pumps, etc).
I wish to get the data it has over MQTT to reach to a HA machine.
The solution about I've heard is to have a Wemos D1/D2 serial connected with the Arduino setup which sends periodically whatever variables I want to have on HA.
I remember I found one single place where a library was mentioned to be added on the Arduino project and configured properly (but I did not saved that page). Probably is about TasmotaSlave.
I mention my available space on the Nano is low (~2k) but probably I can optimize some things to gain more.
A lot of the memory space is probably occupied by the menu screens (i2c display 20x4) 

So my automation has:
- 13 temperatures
- 6 status relays, 2 input status
- total running times for pumps, total MCU running time, time from the last reboot.
- other calculated values
- ~20 configurable parameters

If I would succeed to publish the temperatures and the relay statuses over MQTT, that would be a satisfactory result.
Being able to modify the parameters from HA would be a bonus.

Thank you in advance for any hint toward the solution.

**Comment by barbudor** [comment]
TasmotaClient could certainly help you
https://tasmota.github.io/docs/TasmotaClient/

But why not doing everything with Tasmota?
Why having the Arduino?

**Reply by radudorurus** [reply]
1. I am not sure I will be able to rewrite all the logic behind with multiple conditions using RULES (I have multiple AND/OR combinations for decisions). What about the parameters? I want them displayed directly on a screen not digging through RULES (maybe there is another solution I did not think about). Also, the temperatures need to be on some physical screen not  only the web interface.
  I have considered a migration to ESP8266 without tasmota. That would be a solution where to add only MQTT support and there is.

2. I do not trust wifi and the setup HA_server-tasmota_client for safety stuff. It has to work normal standalone without any third party intervention. I already have issues with sonoffs-tasmota getting out of network and needs router restart to be again in the network. That is the reason I bought ETH01 to migrate an ESP32 power diverter on it having wired LAN connection.

3. The Arduino setup proved quite reliable for 6 years so far. I had only one crash (the Arduino clone become unresponsive, if needed it can be solved by external watchdog reset)

Thanks. I have read the TasmotaClient article but It does not respond to my need. I do not need to flash anything OTA.
Maybe I missed something. Is it about TasmotaSlave > SlaveRespondTele (https://github.com/andrethomas/TasmotaClient/wiki)?
Closer to the scope would be https://tasmota.github.io/docs/Rules/#processing-json-received-from-softwareserialbridge but I would have to manipulate RULES in order to parse the JSONS (I try avoid that complication). I do not need Tasmota to take any decisions, just to relay the data over MQTT.
Or... if the data parsed from serial-JSON would be on the Tasmota webpage and further published over MQTT, that would be a working solution.

**Comment by lalo-uy** [comment]
I don’t know where you run HA, but ther is a daemon for raspberry that bridges serial to mqtt.
https://github.com/vortex314/serial2mqtt

**Reply by radudorurus** [reply]
Thanks.
I considered this solution also.
I have ~4m distance between the arduino and the HA machine.
Probably, a 3-5m usb exttension + USB-TTL 1m cable would do the job to avoid long ttl-uart cables prone to interferences.

First I need to figure out what to send over serial comm (the syntax)
